tsl/src/compression/deltadelta_impl.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+/*
+ * Decompress the entire batch of deltadelta-compressed rows into an Arrow array.
+ * Specialized for each supported data type.
+ */
+
+#define FUNCTION_NAME_HELPER(X, Y) X##_##Y
+#define FUNCTION_NAME(X, Y) FUNCTION_NAME_HELPER(X, Y)
+
+static ArrowArray *
+FUNCTION_NAME(delta_delta_decompress_all, ELEMENT_TYPE)(Datum compressed)
+{
+	StringInfoData si = { .data = DatumGetPointer(compressed), .len = VARSIZE(compressed) };
+	DeltaDeltaCompressed *header = consumeCompressedData(&si, sizeof(DeltaDeltaCompressed));
+	Simple8bRleSerialized *deltas_compressed = bytes_deserialize_simple8b_and_advance(&si);
+
+	const bool has_nulls = header->has_nulls == 1;
+
+	Assert(header->has_nulls == 0 || header->has_nulls == 1);
+
+	/* Can't use element type here because of zig-zag encoding. */
+	uint64 *deltas_zigzag;
+	const int16 num_deltas = simple8brle_decompress_all_uint64(deltas_compressed, &deltas_zigzag);
+
+	Simple8bRleBitmap nulls = { 0 };
+	if (has_nulls)
+	{
+		Simple8bRleSerialized *nulls_compressed = bytes_deserialize_simple8b_and_advance(&si);
+		nulls = simple8brle_bitmap_decompress(nulls_compressed);
+	}
+
+	const int n_total = has_nulls ? nulls.num_elements : num_deltas;
+	const int n_total_padded =
+		((n_total * sizeof(ELEMENT_TYPE) + 63) / 64) * 64 / sizeof(ELEMENT_TYPE);
+	const int n_notnull = num_deltas;
+	const int n_notnull_padded =
+		((n_notnull * sizeof(ELEMENT_TYPE) + 63) / 64) * 64 / sizeof(ELEMENT_TYPE);
+	Assert(n_total_padded >= n_total);
+	Assert(n_notnull_padded >= n_notnull);
+	Assert(n_total >= n_notnull);
+	Assert(n_total <= GLOBAL_MAX_ROWS_PER_COMPRESSION);
+
+	const int validity_bitmap_bytes = sizeof(uint64) * ((n_total + 64 - 1) / 64);
+	uint64 *restrict validity_bitmap = palloc(validity_bitmap_bytes);
+	ELEMENT_TYPE *restrict decompressed_values = palloc(sizeof(ELEMENT_TYPE) * n_total_padded);
+
+	/* Now fill the data w/o nulls. */
+	ELEMENT_TYPE current_delta = 0;
+	ELEMENT_TYPE current_element = 0;
+	uint64 *restrict source = deltas_zigzag;
+#define INNER_SIZE 8
+	Assert(n_notnull_padded % INNER_SIZE == 0);
+	for (int outer = 0; outer < n_notnull_padded; outer += INNER_SIZE)
+	{
+		for (int inner = 0; inner < INNER_SIZE; inner++)
+		{
+			/*
```

Comment:
Could you move the comment before the definition of `INNER_SIZE`? 

Perhaps we could gather more detailed information on the efforts made to vectorize this function. I am curious about the methods/implementation details we have already tried.

Reply:
In other functions, you use `INNER_LOOP_SIZE`. Could you unify the name?

Reply:
Renamed the define.

About the methods, delta-delta decompression is just double prefix sum, so I tried looking up a SIMD algorithm for prefix sum and applying it twice, but it didn't give good results. Then I just unrolled the loop manually to hopefully enable some vectorization. It does vectorize the zig-zag decode part, but not the prefix sum part.